Refactor ImportForm as function component and use hooks

amustaque97 opened the pull request:
This PR is a part of issue #22890

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Refactor `ImportForm` as function componenet. All componenet test case are passing.
After changes, it is working as expected.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Open link `http://localhost:8888/wp-admin/edit.php?post_type=wp_block`
2. Create a new file with name `hello-world.json` and paste the below code
```json
{
  "__file": "wp_block",
  "title": "HelloWorld",
  "content": "<!-- wp:heading -->\n<h2 id=\"hello-world-importform\">Hello, World - ImportForm</h2>\n<!-- /wp:heading -->"
}
```
3. On the reusable blocks page, import the saved file `hello-world.json`
4. Verify the success message.
5. Upload non-json file and verify an error message.

## Screenshots <!-- if applicable -->
* In case of invalid file
<img width="700" alt="image" src="https://user-images.githubusercontent.com/21127788/143787816-9252de9d-6a15-4a0a-a1cf-a02bb6505afb.png">

* In case of a successful import
<img width="1126" alt="image" src="https://user-images.githubusercontent.com/21127788/143787834-73d0054b-e66b-4548-90ed-b0065b0566e9.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Code Quality
## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

amustaque97 committed:
Refactor ImportForm as function component and use hooks

Refactor ImportForm as function componenet. All componenet test case are passing.
After changes, it is working as expected.

youknowriad commented on packages/list-reusable-blocks/src/components/import-form/index.js:
```diff
@@ -11,52 +11,47 @@ import { Button, Notice } from '@wordpress/components';
-		this.onChangeFile = this.onChangeFile.bind( this );
-		this.onSubmit = this.onSubmit.bind( this );
-	}
+	let isStillMounted = true;
```
Should this be a "ref" instead?

amustaque97 replied:
Hi @youknowriad, could you please help me to understand why should we implement `ref` instead of the current way? 

For my understanding I created two code sandbox and it turns out both are behaving same. I would personally prefer to go this way because
- First, we're able to achieve the expected outcome.
- Second, `ref` should be avoided as much as possible as mentioned in the official [docs](https://reactjs.org/docs/refs-and-the-dom.html#when-to-use-refs)

> Avoid using refs for anything that can be done declaratively.
- Third, it will add some unnecessary complexity.

**Code Sandbox link**
- Implementation using `ref` - [link to the sandbox](https://codesandbox.io/s/dazzling-driscoll-uj2o3?file=/src/App.js)
- Implementation without `ref` - [link to the sandbox](https://codesandbox.io/s/dazzling-https-m4xzc)

Would love to hear your thoughts! 😄 

Update: Links has been updated. Looks like there was some problems with the link.

youknowriad replied:
It comes down to safety for me, consider the `onSubmit` handler as written today, it is using a new instance on each render, just like isStillMounted, but if tomorrow we decide to memoize it (useCallback or useMemo for instance) for performance reasons, then the `isStillMounted` that it's going to use is going to be the wrong one because it will use the reference that was created when the callback has been first memoized (meaning isStillMounted will never be false)

amustaque97 replied:
I have made the required changes, please take a look at it.

amustaque97 committed:
Remove `isStillMounted` variable

Introduced `formRef` ref since it will be reliable and safe if in future
we decide to memoize it.

youknowriad reviewed (approved):
LGTM 👍 Thanks for the contribution.
